src/OledDisplay.cpp

```diff
@@ -0,0 +1,192 @@
+}
+
+void DisplayClass::loop()
+{
+    if (display_type > 0) {
+        time_now = millis();
+
+        float totalPower = 0;
+        float totalYieldDay = 0;
+        float totalYieldTotal = 0;
+
+        display->clear();
+        display->firstPage();
+
+        do {
+            // actual seconds
+            unsigned long currentMillis = millis();
+
+            // set Contrast of the Display to raise the lifetime
+            display->setContrast(60);
+
+            for (uint8_t i = 0; i < Hoymiles.getNumInverters(); i++) {
+                auto inv = Hoymiles.getInverterByPos(i);
+                if (inv == nullptr) {
+                    continue;
+                }
+                totalPower = inv->Statistics()->getChannelFieldValue(CH0, FLD_PAC);
+                totalYieldDay = inv->Statistics()->getChannelFieldValue(CH0, FLD_YD);
+                totalYieldTotal = inv->Statistics()->getChannelFieldValue(CH0, FLD_YT);
+            }
+
+            //=====> Actual Production
+            if (totalPower > 0) {
+                display->setPowerSave(false);
+                if (totalPower > 999) {
+                    sprintf(fmtText, "%2.1f", (totalPower / 1000));
+                    printText(String(fmtText) + F(" kW"), 1);
+                } else {
+                    sprintf(fmtText, "%3.0f", totalPower);
+                    printText(String(fmtText) + F(" W"), 1);
+                }
+                previousMillis = currentMillis;
+            }
+            //<=======================
+
+            //=====> Offline
+            else {
+                printText(String("offline"), 1);
+                // check if it's time to enter power saving mode
+                if (currentMillis - previousMillis >= (interval * 2)) {
+                    display->setPowerSave(true);
+                    previousMillis = currentMillis;
+                }
+            }
+            //<=======================
+
+            //=====> Today & Total Production
+            sprintf(fmtText, "%5.0f", totalYieldDay);
+            printText(("today: ") + String(fmtText) + (" Wh"), 2);
+            sprintf(fmtText, "%.1f", totalYieldTotal);
+            printText(("total: ") + String(fmtText) + (" kWh"), 3);
+            //<=======================
+
+            //=====> IP or Date-Time Production
+            if (!(mExtra % 10) && NetworkSettings.localIP())
+                printText(NetworkSettings.localIP().toString(), 4);
+            else {
+                // Get current time
+                time_t now = time(nullptr);
+                tm* local = localtime(&now);
+                char buffer[21];
+                strftime(buffer, sizeof(buffer), "%a %d.%m.%Y %H:%M", local);
+                printText(String(buffer), 4);
+            }
+            display->sendBuffer();
+            dispX = 0;
+            dispY = 0;
+        } while (display->nextPage());
+        while (millis() < time_now + period) {
```

Comment:
Now it is blocking again. Have a look at maybe the mqtt publishing (https://github.com/tbnobody/OpenDTU/blob/64afbf89067b011a43c5963c8f864d51000bb2e0/src/MqttHandleInverter.cpp#L45)
This is a non-blocking implementation